rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,451 @@
+			if _, ok := jwt.GetSigningMethod(token.Header["alg"].(string)).(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("error parsing token")
+			}
+			return []byte(signature), nil
+		})
+		if err != nil {
+			errStr := fmt.Sprintf("500 internal server error: %s", err)
+			a.formatter.Text(w, http.StatusInternalServerError, errStr)
+			return
+		}
+		// Validate token claims
+		if token.Claims != nil {
+			if err := token.Claims.Valid(); err != nil {
+				errStr := fmt.Sprintf("401 Unauthorized: %v", err)
+				a.formatter.Text(w, http.StatusUnauthorized, errStr)
+				return
+			}
+		}
+		// Validate token itself
+		if err := a.validateToken(token, req.URL.Path, req.Method); err != nil {
+			errStr := fmt.Sprintf("401 Unauthorized: %v", err)
+			a.formatter.Text(w, http.StatusUnauthorized, errStr)
+			return
+		}
+
+		provider.ServeHTTP(w, req)
+	})
+}
+
+// Register authenticator-wide security handlers
+func (a *authenticator) registerSecurityHandlers() {
+	a.router.HandleFunc(login, a.loginHandler).Methods(http.MethodGet, http.MethodPost)
+	a.router.HandleFunc(authenticate, a.authenticationHandler).Methods(http.MethodPost)
+	a.router.HandleFunc(logout, a.logoutHandler).Methods(http.MethodPost)
+}
+
+// Login handler shows simple page to log in
+func (a *authenticator) loginHandler(w http.ResponseWriter, req *http.Request) {
+	// GET returns login page. Submit redirects to authenticate.
+	if req.Method == http.MethodGet {
+		r := render.New(render.Options{
+			Directory:  "templates",
+			Asset:      Asset,
+			AssetNames: AssetNames,
+		})
+		r.HTML(w, http.StatusOK, "login", nil)
+	} else {
+		// POST decodes provided credentials
+		credentials := &credentials{}
+		decoder := json.NewDecoder(req.Body)
+		err := decoder.Decode(&credentials)
+		if err != nil {
+			errStr := fmt.Sprintf("500 internal server error: failed to decode json: %v", err)
+			a.formatter.Text(w, http.StatusInternalServerError, errStr)
+			return
+		}
+		token, errCode, err := a.getTokenFor(credentials)
+		if err != nil {
+			a.formatter.Text(w, errCode, err.Error())
+			return
+		}
+
+		// Returns token string.
+		a.formatter.Text(w, http.StatusOK, token)
+	}
+}
+
+// Authentication handler verifies credentials from login page (GET) and writes cookie with token
+func (a *authenticator) authenticationHandler(w http.ResponseWriter, req *http.Request) {
+	// Read name and password from the form (if accessed from browser)
+	credentials := &credentials{
+		Username: req.FormValue("name"),
+		Password: req.FormValue("password"),
+	}
+	token, errCode, err := a.getTokenFor(credentials)
+	if err != nil {
+		a.formatter.Text(w, errCode, err.Error())
+		return
+	}
+	// Writes cookie with token.
+	http.SetCookie(w, &http.Cookie{
+		Name:   cookieName,
+		Path:   "/",
+		MaxAge: int(a.expTime.Seconds()),
+		Value:  token,
+		Secure: false,
+	})
+	// Automatically move to index page.
+	target := "http://" + req.Host + "/"
```

Comment:
Why is this redirecting to `http`, what if HTTPS is used? It should be enough to use `"/"` in redirect.